module: make CJS load from ESM loader

aduh95 opened the pull request:
This PR adds support for `source` in the `load` hook when loading `'commonjs'` modules. CJS modules loaded this way will receive a `require` function that's not the standard CJS one. It implements only a subset of the CJS API, and monkey-patching the CJS API will (mostly[^1]) not affect it. Loading a module using that `require` function will not call into the CJS loader but the ESM one – and similarly, calling `require.resolve` will use the ESM `resolve` hook.

[^1]: There are still some dependency to the legacy loader. The long-term goal would be to get rid of those.

<!--
Before submitting a pull request, please read
https://github.com/nodejs/node/blob/HEAD/CONTRIBUTING.md.

Commit message formatting guidelines:
https://github.com/nodejs/node/blob/HEAD/doc/contributing/pull-requests.md#commit-message-guidelines

For code changes:
1. Include tests for any bug fixes or new features.
2. Update documentation if relevant.
3. Ensure that `make -j4 test` (UNIX), or `vcbuild test` (Windows) passes.

Developer's Certificate of Origin 1.1

By making a contribution to this project, I certify that:

(a) The contribution was created in whole or in part by me and I
    have the right to submit it under the open source license
    indicated in the file; or

(b) The contribution is based upon previous work that, to the best
    of my knowledge, is covered under an appropriate open source
    license and I have the right under that license to submit that
    work with modifications, whether created in whole or in part
    by me, under the same open source license (unless I am
    permitted to submit under a different license), as indicated
    in the file; or

(c) The contribution was provided directly to me by some other
    person who certified (a), (b) or (c) and I have not modified
    it.

(d) I understand and agree that this project and the contribution
    are public and that a record of the contribution (including all
    personal information I submit with it, including my sign-off) is
    maintained indefinitely and may be redistributed consistent with
    this project or the open source license(s) involved.
-->

GeoffreyBooth commented on lib/internal/modules/esm/loader.js:
```diff
@@ -319,6 +336,15 @@ class DefaultModuleLoader {
+    defaultLoadSync ??= require('internal/modules/esm/load').defaultLoadSync;
+
+    const result = defaultLoadSync(url, context);
+    result.format = `require-${result?.format}`;
```
This feels awkward. Could we perhaps track the origin of each load as a distinct property, like whether the load originated as a `require` or an `import`?

isaacs replied:
Wouldn't `import` necessarily have an async load?

isaacs replied:
The `?.` is extraneous, though. Either this is returning a null result, or it isn't. If it is, then assigning to `result.format` will fail.

GeoffreyBooth reviewed (commented):
This is really promising! Great work!

I assume the end goal is to eventually remove the current CommonJS loader? And so we would need to figure out how to make the whole “CJS-only” flow entirely sync, so as not to break async hooks tests.

- how to make `require('./file.json')` work if `require` does not have import attributes but the ESM loader enforce a `type: 'json'` attribute to load JSON modules?

I think we should track the origin of each module load: whether it was via `require` or `import`. Then the answer here is simple: if the origin was `require`, look at the file extension. (I assume that’s how the CJS loader does it today.) Basically if the origin is `require`, apply all the existing CJS semantics with extension searching, etc.

- Calling `require(‘./file.cjs’)` shortcut the cjs-module-lexer pass that detects CJS exports (because WASM instantiation is  async), but if we do a later `import(‘./test.cjs’)` dynamic import gets the expected list of exports?

I feel like we should be able to get away with not needing to know the list of named exports for a CommonJS module. The CJS loader doesn’t know or use them, I think? I know the ESM loader expects them as it does the linking phase as part of inserting ESM modules into V8, but maybe there could be some alternate code path within that part of the flow where if the origin was `require`, to insert the module into V8 the way the CJS loader does now, without the linking phase.

- How to make dynamic import work from CJS?

Well it already works in the current CJS loader; what’s the issue specifically?

- How much of the legacy CJS API do we want to port?

As in, stuff like `require.cache` and `require.resolve` and so on? I assume we’d want to port as much as possible, especially stuff we see used in big popular projects.

- + all the TODOs and probably more

aduh95 commented:
> * Calling `require(‘./file.cjs’)` shortcut the cjs-module-lexer pass that detects CJS exports (because WASM instantiation is  async), but if we do a later `import(‘./test.cjs’)` dynamic import gets the expected list of exports?
> 
> I feel like we should be able to get away with not needing to know the list of named exports for a CommonJS module.

Specifically, here's the use case:

```js
// target.cjs
exports.test = 1;
```

```js
// middlepoint.cjs
require('./target.cjs');
```

```js
// middlepoint.mjs
import { test } from './target.cjs';
```

```js
// entry1.mjs
import './middlepoint.mjs';
import('./middlepoint.cjs');

// No issue, node is able to detect `test` as a named export.
```

```js
// entry1.mjs
import './middlepoint.cjs';
import('./middlepoint.mjs'); // SyntaxError: Named export 'test' not found.
```

Meaning we'll get different results for the same file depending on the order of imports. I don't think that's acceptable, but we might get away with it as long as this is opt-in and experimental.

> Well it already works in the current CJS loader; what’s the issue specifically?

```
import('data:text/javascript,');
^

TypeError: Invalid host defined options
    at Object.eval (eval at loadCJS (node:internal/modules/esm/translators:148:27), <anonymous>:4:1)
    at loadCJS (node:internal/modules/esm/translators:168:3)
    at ModuleWrap.<anonymous> (node:internal/modules/esm/translators:195:21)
    at ModuleJob.run (node:internal/modules/esm/module_job:213:25)

Node.js v21.0.0-pre
```

ljharb commented:
Surely in order to drop the current CJS loading code, every single use case, with no exceptions, would need to be met? The CJS system has been marked "stable" and "locked", I believe, for a very long time, and to me that means that no feature would ever be removed.

GeoffreyBooth commented:
> `import('./middlepoint.mjs'); // SyntaxError: Named export 'test' not found.`

I’m not following this. This is an ESM file importing an ESM file; I would expect it to work. Is this issue that the stack is `entry1.mjs` » `middlepoint.mjs` » `target.cjs`, but since `target.cjs` has been `require`‘d already it’s already in the cache, and therefore it doesn’t go through the whole module importing machinery to find its named exports? Wouldn’t the solution there to be something like marking it on initial `require` as a CJS import that hasn’t had the lexer run on it yet, and then when the later `import` happens it sees that flag and runs the lexer to add that additional metadata? Then the async wasm stuff still only happens in the esm import flow, when we need it; and not on the sync require flow when we can’t do it.

> `TypeError: Invalid host defined options`

This just seems like something to be fixed in the code. Surely there’s some way to make this work in a unified-loader architecture since it already works in the current split loader setup.

aduh95 commented:
> Is this issue that the stack is `entry1.mjs` » `middlepoint.mjs` » `target.cjs`, but since `target.cjs` has been `require`‘d already it’s already in the cache, and therefore it doesn’t go through the whole module importing machinery to find its named exports?

That's my understanding as well.



> This just seems like something to be fixed in the code

That seems obvious, where else would we fix it?

GeoffreyBooth commented:
> That seems obvious, where else would we fix it?

I just mean that this isn't a design question, it's just a matter of fixing a bug. As in we don't need to contemplate changing the API or anything like that.

aduh95 commented:
> As in, stuff like `require.cache` and `require.resolve` and so on? I assume we’d want to port as much as possible, especially stuff we see used in big popular projects.

Speaking of `require.cache`, we have to decide how to fill in `__filename` and `__dirname` for modules that don't come from the FS. I assume that for modules whose URL is a `file:` URL, we can simply use `fileURLToPath`, but for any other protocol, it's going to be an issue I think. `require.cache` keys are paths, so it's the same issue – or we could decide to use a slightly different API, e.g. using URL strings instead of paths as keys for not-`file:` modules.
The only place I could find where a CJS module is loaded by Node.js is for stdin eval, or `--eval` CLI flag, here's how it works:

```console
$ node -p __filename
[eval]
$ node -p __dirname 
.
$ echo 'console.log({__filename,__dirname})' | node
{ __filename: '[stdin]', __dirname: '.' }
```

We could look at how other tools in the exosystem that hook into the CJS loader to provide virtual modules capabilities, but it seems they usually rely on also overriding the fs module so it's transparent to users, but I have a feeling that's not a goal for us.

isaacs commented:
> We could look at how other tools in the exosystem that hook into the CJS loader to provide virtual modules capabilities, but it seems they usually rely on also overriding the fs module so it's transparent to users, but I have a feeling that's not a goal for us.

That's my understanding as well. Node's classic module system has always assumed that the module identifier is a filename, and this matches `__filename`, with the only exceptions being node's `[stdin]` and `[eval]`. There have been other non-Node.js CJS systems that used different semantics, but `__filename` was always a node-ism.

I think a reasonably un-surprising choice would be for `__filename` to be set to the module identifier (eg, `data:text/javascript;...`) if `fileURLToPath(url)` doesn't make sense. The only case I could imagine that causing issues is if we want to do something like `require('https://unpkg.org/express')` some day. But since ESM is already fit for that purpose, I think it'd even be fine for the CJS loader to just not accept anything except `file://` urls.

isaacs commented on lib/internal/modules/esm/translators.js:
```diff
@@ -211,6 +221,7 @@ translators.set('commonjs', async function commonjsStrategy(url, source,
 });
 
 function cjsPreparseModuleExports(filename, source) {
+  // TODO: Do we want to keep hitting the user mutable CJS loader here?
```
I think we need to in order to interoperate with transpiling workflows that overwrite require.extensions today, correct?

aduh95 replied:
I think there would be value in proposing an alternative that gets rid of all the legacy stuff (including `require.extensions`) – even if that can never be the default, because of backward compat – otherwise it's simply too difficult to integrate. Maybe it's a bad idea, we'll see, I'm not married to it, but I'd like us to try it.

isaacs replied:
"Never" is quite a long time. Having a "no legacy" opt in would definitely be a very nice first step towards removing require.extensions eventually.

aduh95 commented:
There are 13 test failures to address:

```
Failed tests:
out/Release/node ./test/parallel/test-node-output-errors.mjs
out/Release/node ./test/parallel/test-runner-cli.js
out/Release/node ./test/es-module/test-esm-cjs-exports.js
out/Release/node --pending-deprecation ./test/es-module/test-esm-exports-deprecations.mjs
out/Release/node ./test/es-module/test-esm-exports.mjs
out/Release/node ./test/es-module/test-esm-imports.mjs
out/Release/node --pending-deprecation ./test/es-module/test-esm-imports-deprecations.mjs
out/Release/node ./test/es-module/test-esm-json-cache.mjs
out/Release/node --loader ./test/fixtures/es-module-loaders/hook-resolve-type.mjs ./test/es-module/test-esm-loader-resolve-type.mjs
out/Release/node ./test/es-module/test-esm-loader-hooks.mjs
out/Release/node ./test/es-module/test-esm-snapshot.mjs
out/Release/node --experimental-network-imports --dns-result-order=ipv4first ./test/es-module/test-http-imports.mjs
out/Release/node ./test/es-module/test-loaders-workers-spawned.mjs
```

There are probably tons of edge cases to also take into consideration – although supporting every CJS edge cases is probably not a goal, but we should at least document what's working and what's not – thankfully @isaacs has volunteered to help with writing tests for that. Any additional help would be appreciated, please consider getting involved :)

ljharb commented:
@aduh95 why is that not a goal? The CJS module loading system is the most stable one in node, which includes all the edge cases.

aduh95 commented:
It's not a goal __for this PR__, I should have say (edited my message above). The long term plans of what to support and how is still open to discussion, the goal in this PR is to achieve a MVP.

isaacs commented:
@aduh95 found a few issues: https://gist.github.com/isaacs/a9aacc1efeca7d565f2f7828467d9e6e

1. The default load is not performed if you just `return nextLoader(spec, context)` in the load method. I'd expect that doing that would proceed to do the "normal" cjs load process, but it seems that with this change, you _must_ return a shortcircuit from a loader for a `"format":"commonjs"` if a loader is present.
2. `require.main` is not set. That should be a reference to the `module` object for the first module loaded via process.argv[1].
3. Using the default resolver via `return nextResolve(spec, context)` works for the first main module, but raises an error for subsequent modules loaded via `require()`.

aduh95 commented:
> 1. The default load is not performed

It is performed, the issue was that the new translator didn't support `ArrayBuffer` sources, but that's now fixed. I've also fixed the other points you mentioned, we're now at 10 test failures.

Qard commented:
Can you please wait for a review from me before landing this? The off-thread loader change broke ESM support in APMs and we want to make sure this doesn't break CJS too.

GeoffreyBooth commented:
> Can you please wait for a review from me before landing this? The off-thread loader change broke ESM support in APMs and we want to make sure this doesn’t break CJS too.

Can you make a list of APIs that are necessary to preserve?

I would imagine that APMs will just be required to start using the loaders API, even for CommonJS, to preserve current functionality. We’ll probably land #46826 before this (or we can ensure that it’s released before this) which should make the upgrade path simpler for APM authors and hidden from end users; the `--require apm-name` or `require('apm-name')` calls that users are making now could call `register` to register hooks with the loaders API. This has the pleasant side effect (from Node’s perspective) of forcing APM vendors to catch up with Node and add full support for ESM apps.

Qard commented:
Forcing APMs to "catch up" is going to go very badly. ESM changes way too quickly for APMs to keep up with it unless we have people full time on just that and that's not likely to happen. We're more likely to just not support those Node.js versions and push harder on increasing diagnostics_channel adoption. Most APM users are very slow to upgrade anyway.

JakobJingleheimer commented:
I'm a little confused. Off-threading was over ½ a year in the making, and the design from an external perspective never changed throughout that period. Why is it a game of catch-up now, and what do you need to make that easier? (If this is a lengthy discussion, let's move it to the loaders repo).

PS We are very, very eager to know use-cases in the wild we need to consider, and we're actively collecting them. If you can detail your considerations, that would be most welcome—or, better yet, codify them in tests it make it obvious if we've broken them.

GeoffreyBooth commented:
@Qard Please provide a list of CommonJS APIs that APMs rely on and we can try to preserve their functionality while getting the Loaders API to support CommonJS. But in my mind, if there’s any conflict between those goals, finishing the Loaders API takes priority. It’s just broken without full CommonJS support, and we can’t leave it in that state indefinitely just because some vendors who are profiting off of Node won’t invest in supporting Node’s newest versions. That said, I’m not _trying_ to break anyone so if there’s any way to maintain compatibility with monkey-patching CommonJS internals or whatever techniques these tools rely on, I’m happy to continue supporting it whenever possible.

mcollina commented:
> I would imagine that APMs will just be required to start using the loaders API, even for CommonJS, to preserve current functionality.

Can you clarify this point? Would this break how `require` work?
I don't think we can afford to break commonjs backward compat.

arcanis commented:
I agree that adding ESM to CJS should be designed as an extension to be successful, not a replacement.

GeoffreyBooth commented:
> Can you clarify this point?

That was just a hypothetical musing, that we don't need to debate as it's unlikely to happen at least anytime soon. The goal with this PR is to get all tests passing, which would mean we're not breaking any backward compatibility for any CommonJS APIs that are documented. My question for @Qard is whether there are any CommonJS APIs that APMs are relying on that aren't covered by tests; things like monkey patching that we don't officially support, but can try to preserve if this PR breaks them.

Qard commented:
See: https://github.com/DataDog/dd-trace-js/blob/master/packages/dd-trace/src/ritm.js

I won't have time to get into more detail until Tuesday, but there are a bunch of underscore-prefixed methods on the Module class which need to be accessible for APMs to continue working.

My plan has been to migrate people to diagnostics_channel so module patching becomes no longer relevant, however that module patching mechanism needs to continue to work for at least the mid-term.

It's important to understand that you're dealing with big enterprises that plan things in quarters or years. If you want to break compatibility with something, especially something this heavily relied upon, you can't just go do that in a few months you generally need a year or more to migrate users and there needs to be something to migrate to before that can happen. And pre-release discussion doesn't count, there's no visibility for that unless a company has people actively engaging in that group which is generally not the case. That's why the off-thread loaders thing was so disruptive, because there was no flag to turn it off in the mid-term. Yes there was some leading work that went into it, but we don't have people actively engaging in the loaders group because we are a small team and don't have the time to be in every discussion. At most APMs the Node.js team is only a couple people and most of their time is going into delivering OKRs. I would love to have people working on Node.js core full-time, and I've been pushing for that, but that's simply not the reality, especially right now with hiring across the market being so tight.

JakobJingleheimer commented:
> there are a bunch of underscore-prefixed methods

That sounds like AMPs are doing something naughty and a well-known no-no 🙃 also, i'm not sure it's reasonable to ask the millions of people (many of who probably are not even using AMPs) to wait months or years for AMPs to update.

Off-threading was on a semver release and not back-ported to older lines, so it didn't break anything—it just prevented slower consumers upgrading, which is what happens all the time for anything (take, for example, my own day-job where we use ESBuild: like every release is a breaking change and we don't have time every other week to deal with that, so we lag a couple minors behind).

That said, I would definitely prefer to not break anybody's stuff, so let's take a look at what those private-ish methods are doing and see what we can do 🙂

arcanis commented:
> That sounds like AMPs are doing something naughty and a well-known no-no 🙃

I think you're underestimating how widespread patching is in the ecosystem to compensate Node.js' limitations. APMs do it, but they're far from alone. Yarn does it. Babel does it. Ts-node does it. Jest as well. Anything using pirates does.

A proper resolution API should have been provided a long time ago, but it hasn't (in part because ESM and loaders were slowly progressing, making it difficult to suggest parallel improvements on the CJS pipeline) and now I think Node has to "pay the price", so to speak.

That's not to say that integrating loaders in the CJS resolution is a bad idea, but I think "let's look at what we can keep" is the wrong approach - you shouldn't need to keep this list, because you shouldn't break the primitives (what I mentioned by extension over replacement).

aduh95 commented:
I agree we should not try to make a list, and instead build a system that’s monkey-patching free. Hopefully the ESM loader hooks API provides a cleaner API that is as powerful (or maybe even more so) as the CJS loader monkey-patching. Of course that can only work if the changes are opt-in (e.g. using a CLI flag), and we can switch the default only if the ecosystem follows along – and if the ecosystem doesn’t follow, so be it, we keep supporting monkey-patching, but that’d be a failure imho.

JakobJingleheimer commented on lib/internal/modules/esm/loader.js:
```diff
@@ -365,9 +396,17 @@ class CustomizedModuleLoader extends DefaultModuleLoader {
+    return this.getJobFromResolveResult(resolveResult, parentURL, importAssertions, false);
   }
-  getModuleJob(specifier, parentURL, importAssertions) {
+  #getModuleJobSync(specifier, parentURL, importAssertions) {
```
nit

```suggestion

  #getModuleJobSync(specifier, parentURL, importAssertions) {
```

JakobJingleheimer commented on lib/internal/modules/esm/loader.js:
```diff
@@ -365,9 +396,17 @@ class CustomizedModuleLoader extends DefaultModuleLoader {
+
+    return this.getJobFromResolveResult(resolveResult, parentURL, importAssertions, true);
+  }
+  getModuleJob(specifier, parentURL, importAssertions, sync) {
```
nit

```suggestion

  getModuleJob(specifier, parentURL, importAssertions, sync) {
```

JakobJingleheimer commented on lib/internal/modules/esm/loader.js:
```diff
@@ -393,6 +432,17 @@ class CustomizedModuleLoader extends DefaultModuleLoader {
 
+    return result;
+  }
+  loadSync(url, context) {
```
nit

```suggestion

  loadSync(url, context) {
```

JakobJingleheimer replied:
Huh? this is already defined on like 340, no?

GeoffreyBooth commented:
Can you please start a discussion thread to debate these issues? And leave this thread to discuss this particular PR.

Qard commented:
> That sounds like AMPs are doing something naughty and a well-known no-no 🙃

Similar to req._headers, this has been a known issue for a long time. Core has for many years rejected suggestions to make these things we need public and APMs have unfortunately not got the management buy-in to contribute some better system there. I'm with you that APMs should be more involved, but cutting them out would be a highly damaging move due to how widespread their use is. Literally 90% of all apps out there depend on these module internals somewhere in their dependency graph, either through APMs or other things. (I know, I've checked.) Node.js was always very cautious about breaking widespread use-cases like that until the recent turbulence from loaders. We can't push through changes that will break so many users without some sort of mitigation path in-place.

@GeoffreyBooth I don't agree this should be in another thread. This discussion is directly related to the outcome of landing this PR specifically.

GeoffreyBooth commented:
>  I don't agree this should be in another thread. This discussion is directly related to the outcome of landing this PR specifically.

This PR can land or not contingent on the discussion that happens elsewhere. Please don't continue posting here with theoretical concerns. Let's leave the discussion here related to specific lines of code in this PR.

mcollina commented:
I find GitHub discussions extremely hard to collaborate on. Please keep the technical discussion in here or on an issue.

mcollina commented:
Tagging tsc-agenda for visibility.

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -169,7 +169,7 @@ class DefaultModuleLoader {
     }
 
     if (job === undefined) {
-      job = this.#createModuleJob(url, resolvedImportAssertions, parentURL, format);
+      job = this.#createModuleJob(url, resolvedImportAssertions, parentURL, format, sync);
     }
```
`job === undefined` is not the same thing as `job == null`

JakobJingleheimer replied:
`null` is not a valid value of `job`, is it? (in which case, it's a distinction without a difference—it doesn't matter)

GeoffreyBooth commented:
I just checked out this branch to see where it’s at. On my machine, I have 10 failures, many of which appear related:

- `./node ./node/test/parallel/test-node-output-errors.mjs`
- `./node --experimental-loader ./test/fixtures/es-module-loaders/assertionless-json-import.mjs ./node/test/es-module/test-esm-assertionless-json-import.js`
- `./node ./node/test/es-module/test-esm-cjs-exports.js`
- `./node --pending-deprecation ./node/test/es-module/test-esm-exports-deprecations.mjs`
- `./node ./node/test/es-module/test-esm-exports.mjs`
- `./node ./node/test/es-module/test-esm-imports.mjs`
- `./node --pending-deprecation ./node/test/es-module/test-esm-imports-deprecations.mjs`
- `./node ./node/test/es-module/test-esm-json-cache.mjs`
- `./node ./node/test/es-module/test-esm-loader-hooks.mjs`
- `./node ./node/test/es-module/test-esm-snapshot.mjs`

So it’s probably not far from being ready. And we do happen to have [one test for monkey-patching CommonJS](https://github.com/nodejs/node/blob/main/test/parallel/test-module-run-main-monkey-patch.js) in the codebase, which passes, so at least some (possibly all) monkey-patching of CommonJS remains working with this PR.

Qard commented:
Yep, I did some manual testing against the Datadog tracer today and it's still working. At least as much as ESM was working before the change. And from what I can tell this would only impact CJS loaded with import and not CJS loaded with require which is our _main_ concern. I'm less worried about breakage from this particular PR now, but I'll give a full code review when it's ready. 🙂

Sorry if I'm being a bit over-cautious, but changes to the module system are a significant existential threat the the continued functioning of APM vendors, so we _really_ don't want to risk problems there. For ESM support for that is a feature request, but if CJS breaks that's a major bug for us and would impact a whole lot of users. We just want to be sure things aren't going to catch fire suddenly. 😅

targos commented:
Yeah, from my PoV, the problem is not really that APMs and other tools rely on undocumented/legacy stuff. It's that we have almost no test coverage for it.
If a use case is broken because we had no tests to tell us that we are breaking it before landing a change, there's not much we can do.

Qard commented:
> > Yep, I did some manual testing against the Datadog tracer today and it's still working.
> 
> Can you share steps that we could follow to do the same test? That would be a good thing to be able to do, as a way to know if we're likely to break instrumentation packages.

It's a fairly manual process at the moment and needs a Datadog account. I'm hoping to find some time soon to make sure the relevant bits of ritm and iitm are present in the test suite so we can avoid any future breakage. From what I understand there's already at least _parts_ of what ritm in the test suite, but I need to evaluate that and make sure all the parts we need are there.

aduh95 commented:
@isaacs now that tests are passing on most platforms, any chance you'll be able to test that again?

aduh95 commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -186,6 +195,21 @@ class ModuleJob {
+      return { __proto__: null, module: this.module };
+    }
+
+    // TODO: better handle errors
```
```suggestion
```

GeoffreyBooth reviewed (commented):
This is very, very impressive. I didn’t dare to dream that it would be possible to pipe `require` calls through the Loaders API, but not only is it possible but it’s relatively easy, as this PR proves. Congratulations!

Should we add some text to the docs somewhere explaining a bit more about this? I know we add a note about returning `format: 'commonjs'` but I think it’s a bit higher level that the Loaders API can support `require` now, and in order to take advantage of this capability one only needs to return `source` for `format: 'commonjs'`. Perhaps in the introduction to the Loaders section? We should also mention that for any loaders that do opt in, both `import` and `require` calls in those CommonJS modules are handled via the custom hooks.

Is the opt-in as easy as:
- Loader starts returning `source` for `format: 'commonjs'`
- User registers that loader, such as via `--import ts-node/register` say

I assume it is, because using `--import` or `--loader` means that the entry point is handled by the ESM loader, and then since this opt-in loader is registered, even if the entry point is CommonJS it’ll get handled by the registered hooks. So really the only way to opt _out_ while using a loader would be to run via `node --require ts-node/register.cjs entry.cjs` where both the loader registration and the entry point are processed by the CommonJS loader. We should probably find a way to make `register()` calls opt into the ESM loader being used, like how `--import` or `--loader` opt into it, to avoid confusing behavior. This can be a follow-up, but it would be good to document somewhere (maybe just in a comment on this PR) what the behavior us as of this PR. Or maybe add a test like `node entry.cjs` where the entry is `register('/path/to/loader-fixture.mjs'); require('/path/to/file-affected-by-loader-hooks-since-registration-is-sync.cjs')`, and if it “fails” currently the test can note that it’s a TODO to make it pass in the future.

It also feels like we should have more than just two modified tests as a result of these changes. Is there a test that validates the happy path, that CommonJS `require` calls are handled by the registered hooks? That both `resolve` and `load` get called for `require`? That children of CommonJS modules get handled via the same flow? That this all happens when the entry point is a CommonJS file, at least when the loader is registered via `--import`? Also we should add some tests around monkey-patching like @isaacs mentioned, both for current (before this PR) behavior and also to document expected behavior under the opt-in conditions of this PR (which I assume is just, “monkey patching no longer works”).

Again, great work here. I can’t wait to get this landed!

mcollina commented:
Do you plan to ship this as a minor or a major?

JakobJingleheimer commented:
> Do you plan to ship this as a minor or a major?

I believe we decided last meeting to somehow make it opt-in for a while first.

aduh95 commented on lib/internal/modules/esm/load.js:
```diff
@@ -69,6 +69,33 @@ async function getSource(url, context) {
+    const { readFileSync } = require('fs');
+    source = readFileSync(parsed);
+  } else if (parsed.protocol === 'data:') {
+    const match = RegExpPrototypeExec(DATA_URL_PATTERN, parsed.pathname);
```
What do you mean by resetting the prototype?

GeoffreyBooth replied:
Resetting the internal state that gets stored in regex. There was a PR about this a few months ago, with tests in the repo.

aduh95 replied:
You're probably thinking about regex with the `g` flag, or something like that, that doesn't apply here.

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -385,6 +409,23 @@ class ModuleLoader {
+      defaultLoadSync(url, context);
+    let format = result?.format;
+    if (format === 'commonjs') {
+      format = 'require-commonjs';
```
It's the `loadSync` method, that code is executed only subsequently to a `require()` call.

aduh95 commented on lib/internal/modules/esm/translators.js:
```diff
@@ -184,30 +254,58 @@ translators.set('commonjs', async function commonjsStrategy(url, source,
+}
+
+translators.set('require-commonjs', (url, source, isMain) => {
+  assert(cjsParse);
+
+  return createCJSModuleWrap(url, source);
 });
 
-function cjsPreparseModuleExports(filename) {
+translators.set('commonjs', async function commonjsStrategy(url, source,
```
> do we need two translators for CommonJS?

One is sync, the other one is async. Yes I think we need two of them.



> Is one `commonjs-via-commonjs-loader` and the other `commonjs-via-esm-loader`?

The CJS loader never calls into the ESM one, so all calls to those translator originates from the ESM one.

> Or is one `commonjs-via-require` and the other `commonjs-via-import`?

Yeah, well the async one covers more than just import (e.g. the entry point is also loaded async).

GeoffreyBooth replied:
How about `commonjs-sync` and `commonjs-async`?

aduh95 replied:
Well CJS is never async, that could be confusing. That being said, this is internal only, so I'm not going to fight for it, I'm happy to use whatever.

GeoffreyBooth replied:
I suggested those names because you wrote “One is sync, the other one is async.” Yet you also say that CJS is never async. So maybe you should explain in more detail the differences between these two translators and when each is used.

aduh95 replied:
One translator is async, the other one is sync. CJS modules are sync, in the sense they do not support TLA.

GeoffreyBooth replied:
> The other one is used for all the other cases when we want to load CJS.

Those other use cases are just `import` or the entry point, right? Assuming the entry is being handled by the ESM loader because of `--import` or `--loader` usage. So maybe just `require-commonjs` and `import-commonjs`, with explanatory comments:

```js
// Handle CommonJS modules referenced by `require` calls.
// This translator function must be sync, as `require` is sync.
translators.set('require-commonjs', (url, source, isMain) => {
  assert(cjsParse);

  return createCJSModuleWrap(url, source);
});

// Handle CommonJS modules referenced by `import` statements or expressions,
// or as the initial entry point when the ESM loader handles a CommonJS entry.
translators.set('import-commonjs', async function commonjsStrategy(url, source,
```

aduh95 commented on lib/internal/modules/esm/translators.js:
```diff
@@ -219,38 +317,43 @@ function cjsPreparseModuleExports(filename) {
+      const reexport = reexports[i];
+      let resolved;
+      try {
+        // TODO: this should be calling the `resolve` hook chain instead.
+        resolved = CJSModule._resolveFilename(reexport, module);
```
It means the re-export of CJS loader will not always have the "correct" exported names – which would be good to fix, but I doubt it's going to block anyone from using the feature. I'm not going to fix it in this PR, feel free to do it (in this PR or in a follow up).

aduh95 replied:
🧐 this is the lexer step, what do you mean? The lexer step is relying on the CJS loader, which is unfortunate, but that needs to be the case for now (because it must be sync, so that would mean removing support for CJS in the loader thread, which I don't want to in this PR)

GeoffreyBooth replied:
Let’s back up a bit. I’m not really following what this comment is meaning or your first explanation of it. Like I don’t know what “the re-export of the CJS loader will not have the correct exported names” has to do with not calling the `resolve` hooks. Do you mind explaining this a bit more? And maybe the comment should be a bit more detailed.

aduh95 commented on lib/internal/modules/esm/translators.js:
```diff
@@ -219,38 +317,43 @@ function cjsPreparseModuleExports(filename) {
+      } catch {
+        continue;
+      }
+      // TODO: this should be calling the `load` hook chain and check if it returns
+      // `format: 'commonjs'` instead of relying on file extensions.
+      const ext = extname(resolved);
+      if ((ext === '.js' || ext === '.cjs' || !CJSModule._extensions[ext]) &&
+          isAbsolute(resolved)) {
```
I don't think so, this code will not be run unless a loader returned a non-nullish source for a CJS module. Same thing, I'm not going to fix it, you're welcome to push something to my branch so if you feel strongly.

aduh95 replied:
Yeah but it's the lexer step, it's not a big deal

aduh95 commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -186,6 +195,20 @@ class ModuleJob {
   }
 
+  runSync() {
+    assert(this.module instanceof ModuleWrap);
```
Why do we need any assertion?

GeoffreyBooth replied:
It's a sincere question. Is this here because you were debugging and forgot to delete it, or are there cases where this throws? If the latter, what cases? Can you add a comment here mentioning what they are?

aduh95 replied:
> If the latter, what cases?

Like the error message says, it can be caused "by either a bug in Node.js or incorrect usage of Node.js internals." Obviously it should never happen, but that's the point of having an assertion. I don't think a comment is necessary.

GeoffreyBooth replied:
If there are no cases when it can happen, other than us introducing bugs into the code, then it should probably be a test rather than something that runs on runtime for every module. This is just an unnecessary line of code that slows things down.

aduh95 replied:
Couldn't that be said about any other assertion in the codebase?

GeoffreyBooth replied:
Yeah, probably. Presumably that’s why there are very few assertions in the codebase.

aduh95 commented on lib/internal/modules/esm/load.js:
```diff
@@ -69,6 +69,33 @@ async function getSource(url, context) {
+    source = BufferFrom(decodeURIComponent(body), base64 ? 'base64' : 'utf8');
+  } else {
+    const supportedSchemes = ['file', 'data'];
+    if (experimentalNetworkImports) {
+      ArrayPrototypePush(supportedSchemes, 'http', 'https');
+    }
```
```suggestion
```

aduh95 commented on lib/internal/modules/esm/load.js:
```diff
@@ -64,6 +64,33 @@ async function getSource(url, context) {
+    }
+    const { 1: base64, 2: body } = match;
+    source = BufferFrom(decodeURIComponent(body), base64 ? 'base64' : 'utf8');
+  } else {
+    const supportedSchemes = ['file', 'data'];
+    if (experimentalNetworkImports) {
+      ArrayPrototypePush(supportedSchemes, 'http', 'https');
+    }
+    throw new ERR_UNSUPPORTED_ESM_URL_SCHEME(parsed, supportedSchemes);
```
```suggestion
  } else {
    const supportedSchemes = ['file', 'data'];
    throw new ERR_UNSUPPORTED_ESM_URL_SCHEME(parsed, supportedSchemes);
```

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -248,8 +254,25 @@ class ModuleLoader {
+        importAssertions,
+      });
+
+      const translator = getTranslators().get(finalFormat);
```
I copied that from the original implementation. If it's indeed faster, it can be dealt with in a follow-up PR.

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -186,8 +186,25 @@ class DefaultModuleLoader {
-  #createModuleJob(url, importAssertions, parentURL, format) {
-    const moduleProvider = async (url, isMain) => {
+  #createModuleJob(url, importAssertions, parentURL, format, sync) {
+    const moduleProvider = sync ? (url, isMain) => {
+      const {
+        format: finalFormat,
+        responseURL,
+        source,
+      } = this.loadSync(url, {
+        format,
+        importAssertions,
+      });
+
+      const translator = getTranslators().get(finalFormat);
+
+      if (!translator) {
+        throw new ERR_UNKNOWN_MODULE_FORMAT(finalFormat, responseURL);
+      }
+
+      return FunctionPrototypeCall(translator, this, responseURL, source, isMain);
+    } : async (url, isMain) => {
```
One is an async function, the other one is not async. We do need two functions, I'll see if we can improve readability.

aduh95 commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -186,6 +195,20 @@ class ModuleJob {
+
+    this.module.instantiate();
+    this.instantiated = PromiseResolve();
+    const timeout = -1;
+    const breakOnSigint = false;
+    this.module.evaluate(timeout, breakOnSigint);
```
I don't know either, I think it evaluates the module (whatever that means).

GeoffreyBooth replied:
As in, these were lines that existed before, that you just moved?

aduh95 replied:
Ah yes, it's from the run method, just below.

JakobJingleheimer replied:
Ah, I know the answer to this (what this does) from by spelunking into ModuleJob in https://github.com/nodejs/node/pull/48439. I'm gonna do a tidy PR after all this stuff lands to add missing docs etc; I'll add an explanation there.

aduh95 commented on lib/internal/modules/esm/translators.js:
```diff
@@ -140,35 +142,103 @@ function enrichCJSError(err, content, filename) {
-translators.set('commonjs', async function commonjsStrategy(url, source,
-                                                            isMain) {
+/**
+ * Loads a CJS module without using the monkey-patchable CJS loader. Modules
+ * loaded with this method have a special `require` function that calls the ESM
+ * loader.
```
I think I prefer my version.

> Loads a CommonJS module via the ESM Loader.

Well it was already possible to `import` a CJS module via the ESM loader, but it was using the CJS loader under the hood. The important part is that this function provides an alternative that mostly do not depend on the CJS loader.

> Any monkey patches applied to the CommonJS Loader will not affect this module.

That's not (yet) true.

> Any `require` calls in this module will load all children the same way, via the ESM loader.

I think I prefer my version, folks reading the Node.js source will be more interested in the implementation detail than the observable results.

GeoffreyBooth replied:
How about:

```suggestion
 * Loads a CommonJS module via the ESM Loader sync CommonJS translator.
 * This translator creates its own version of the `require` function passed into CommonJS modules.
 * Any monkey patches applied to the CommonJS Loader will not affect this module.
 * Any `require` calls in this module will load all children in the same way.
```

aduh95 commented on lib/internal/modules/esm/translators.js:
```diff
@@ -140,35 +142,103 @@ function enrichCJSError(err, content, filename) {
+  };
+  setOwnProperty(requireFn, 'resolve', function resolve(spec) {
+    if (!StringPrototypeStartsWith(spec, 'node:')) {
+      const path = CJSModule._resolveFilename(spec, module);
+      if (spec !== path) {
+        spec = `${pathToFileURL(path)}`;
+      }
+    }
+    const { url: resolvedURL } = asyncESM.esmLoader.resolveSync(spec, url, kEmptyObject);
+    return StringPrototypeStartsWith(resolvedURL, 'file://') ? fileURLToPath(resolvedURL) : resolvedURL;
+  });
+  setOwnProperty(requireFn, 'cache', CJSModule._cache);
```
I guess we want to deprecate `require.cache`? In a world where the CJS loader is refactored away, `require.cache` cannot be anything useful.

aduh95 commented on lib/internal/modules/esm/translators.js:
```diff
@@ -184,30 +254,58 @@ translators.set('commonjs', async function commonjsStrategy(url, source,
+                                                            isMain) {
+  if (!cjsParse) await initCJSParse();
+
+  // For backward-compatibility, it's possible to return a nullish value for
+  // commonjs source associated with a file: URL. In this case, the source is
+  // obtained by calling the monkey-patchable CJS loader.
```
Hum no, it's nullish values only. Passing a falsy one (e.g. an empty string) will not trigger the backward compatible code path.

aduh95 commented on lib/internal/modules/esm/translators.js:
```diff
@@ -184,30 +254,58 @@ translators.set('commonjs', async function commonjsStrategy(url, source,
       return { module, exportNames: cached.exportNames };
   }
   const loaded = Boolean(module);
+  debug(`IsLoaded CJSModule ${filename}? ${loaded}`);
```
```suggestion
```

GeoffreyBooth commented on doc/api/esm.md:
```diff
@@ -879,20 +882,20 @@ validating the import assertion.
-does not provide a mechanism for the ES module loader to override the
-[CommonJS module return value](#commonjs-namespaces). This limitation might be
-overcome in the future.
+not possible to replace the value of a Node.js builtin (core) module. Unless a
+non-nullish `source` is supplied for type `'commonjs'`, it will be fetched using
+the CommonJS module loader. When non-nullish `source` is supplied for type
+`'commonjs'`, `require` calls from inside the loaded module will call the
+`resolve` and `load` hooks.
```
Okay. What _is_ that behavior? 😀

aduh95 replied:
`defaultLoad` returns `null` as a source when `format` is `'commonjs'`.

GeoffreyBooth replied:
> `defaultLoad` returns `null` as a source when `format` is `'commonjs'`.

So if someone is trying to use the return value of `next`, when `next` is `defaultLoad`, would get a `null` source? Shouldn’t they get the actual source content?

aduh95 replied:
They'll get a `null` source, they would need to call `readFile(new URL(url))` to opt-in to the new behavior.

GeoffreyBooth replied:
Okay, how about:


```suggestion
not possible to replace the value of a Node.js builtin (core) module.

For type `'commonjs'`, if `source` is defined, `require` calls from inside the
loaded module will go through the registered `resolve` and `load` hooks.
If `source` is undefined or `null`, it will be handled by the CommonJS
module loader and not processed by any registered hooks. This behavior
for nullish `source` is temporary; in the future, a `source` value that is
undefined or `null` will not be supported.

The Node.js internal `defaultLoad`, which is the value of `next` for the last
loader in the `load` chain, does not return a value for `source`. If one is desired,
loader authors should load it via a call like `readFile(new URL(url))`. This may
also change once support for nullish `source` is discontinued.
```

GeoffreyBooth commented on lib/internal/modules/esm/loader.js:
```diff
@@ -267,6 +264,15 @@ class ModuleLoader {
 
       return FunctionPrototypeCall(translator, this, responseURL, source, isMain);
     };
+    const moduleProvider = sync ?
+      (url, isMain) => callTranslator(this.loadSync(url, {
+        format,
+        importAssertions,
+      }), isMain) :
+      async (url, isMain) => callTranslator(await this.load(url, {
+        format,
+        importAssertions,
+      }), isMain);
```
```suggestion
    const context = { format, importAssertions };
    const moduleProvider = sync ?
      (url, isMain) => callTranslator(this.loadSync(url, context), isMain) :
      async (url, isMain) => callTranslator(await this.load(url, context), isMain);
```

GeoffreyBooth commented on doc/api/esm.md:
```diff
@@ -879,20 +882,39 @@ validating the import assertion.
+export async function load(url, context, nextLoad) {
+  const response = await nextLoad(url, context);
+  if (response.format === 'commonjs') {
+    response.source ??= await readFile(new URL(response.responseURL ?? url));
```
I don’t see `responseURL` documented in the signature for `load`, don’t we need to add it?

Qard reviewed (dismissed):
I had our team try out the PR and it does indeed seem to break `import-in-the-middle`. We tried three different app configurations:

<details>
  <summary>ESM entrypoint with import of CJS</summary>

```js
import http from 'http'

const server = http.createServer((req, res) => {
  res.end('done')
}).listen(3000)
```

</details>

<details>
  <summary>CJS entrypoint with require of CJS</summary>

```js
const http = require('http')

const server = http.createServer((req, res) => {
  res.end('done')
}).listen(3000)
```

</details>

<details>
  <summary>CJS with dynamic import of CJS</summary>

```js
import('http').then(http => {
  const server = http.createServer((req, res) => {
    res.end('done')
  }).listen(3000)
})
```

</details>

We run each with `node --loader dd-trace/loader-hook.mjs --require dd-trace/init app.js`.

In Node.js 20.5.0 all variants patch the http module correctly and we get valid traces. With this PR we only get valid traces from CJS loading CJS, so both the dynamic and static import of CJS seems broken. In discussions with @GeoffreyBooth the other day it sounds like this _could_ possibly be a bug on the part of IITM which only surfaces because of the change here, but it _does_ mean that landing this will currently break us. I would like to block landing this until the source of the failure can be confirmed and we know what needs to be done to fix it.

aduh95 commented:
@Qard does `http` in your examples references the built-in `node:http` module or some CJS module via a `resolve` hook? If it's the former, FYI that's not CJS, built-in modules have their own translator separate from CJS.

Qard commented:
It'll have to wait until tomorrow. The person working on that is offline for the day.

Qard commented:
Okay, got confirmation from my teammate that it is in fact broken in main _before_ this change so I'll dismiss the request changes. Will need to figure out what broke that though. What else is in main since 20.5 that could cause such problems?

GeoffreyBooth commented:
> What else is in main since 20.5 that could cause such problems?

The recent `initialize` and `register` PRs.

aduh95 commented:
Sorry to hear that, but also relieved that this PR didn’t introduce some breaking change which would have been unexpected. Maybe you (or your teammate) can open an issue with some details on the failure you’re seeing. You could try to revert each commit that touched `lib/internal/modules/esm` until you are no longer seeing a failure, or I can do that for you if you share a repro.

Qard commented:
We'll get to it soon, just busy with other stuff for the next week. It's not super critical though since CJS require of CJS _does_ seem to always work and that was our _main_ concern. ESM loaders being broken in isolation is not going to impact us much as barely anyone uses it currently, so getting _that_ part working is much less urgent.

JakobJingleheimer commented on doc/api/esm.md:
```diff
@@ -945,20 +948,40 @@ validating the import assertion.
-overcome in the future.
+not possible to replace the value of a Node.js builtin (core) module.
+
+For type `'commonjs'`, if `source` is defined, `require` calls from inside the
+loaded module will go through the registered `resolve` and `load` hooks.
+If `source` is undefined or `null`, it will be handled by the CommonJS
+module loader and not processed by any registered hooks. This behavior
+for nullish `source` is temporary; in the future, a `source` value that is
+undefined or `null` will not be supported.
```
nit (the first bit reads a little yoda-esque), and I think it's worth specifically calling out CJSLoader won't be used and require.extensions won't apply.

```suggestion
If a `load` hook returns commonjs `source`, that module's `require` calls will
be processed by ESM loader with registered `resolve` and `load` hooks;
`require.extensions` and monkey-patching on the CommonJS module loader will not
apply. If `source` is undefined or `null`, it will be handled by the CommonJS
module loader and not processed by registered hooks. This behavior for nullish
`source` is temporary—in the future, nullish `source` will not be supported.
```

JakobJingleheimer commented on doc/api/esm.md:
```diff
@@ -945,20 +948,40 @@ validating the import assertion.
+import { readFile } from 'node:fs/promises';
+
+export async function load(url, context, nextLoad) {
+  const response = await nextLoad(url, context);
+  if (response.format === 'commonjs') {
+    response.source ??= await readFile(new URL(response.responseURL ?? url));
+  }
+  return response;
```
nit: the rest of this doc refers to this as "result"

```suggestion
  const result = await nextLoad(url, context);
  if (result.format === 'commonjs') {
    result.source ??= await readFile(new URL(result.responseURL ?? url));
  }
  return result;
```

JakobJingleheimer commented on lib/internal/modules/esm/load.js:
```diff
@@ -69,12 +70,35 @@ async function getSource(url, context) {
 }
 
+function getSourceSync(url, context) {
+  const parsed = new URL(url);
+  const responseURL = url;
+  let source;
+  if (parsed.protocol === 'file:') {
+    source = readFileSync(parsed);
+  } else if (parsed.protocol === 'data:') {
```
`URL::protocol` (and all props not `href`) is computed each read, so we can get a bit of a perf improvement by caching its value (which isn't changed):

```suggestion
  const parsed = new URL(url);
  const { protocol } = parsed;
  const responseURL = url;
  let source;
  if (protocol === 'file:') {
    source = readFileSync(parsed);
  } else if (protocol === 'data:') {
```

also, nit: i think this would be more grokkable as a `switch` because we immediately know all cases are related to `protocol`.

aduh95 replied:
I copied that from `defaultLoad`, we should probably address both in a follow up PR.

JakobJingleheimer commented on lib/internal/modules/esm/load.js:
```diff
@@ -69,12 +70,35 @@ async function getSource(url, context) {
+    const { 1: base64, 2: body } = match;
+    source = BufferFrom(decodeURIComponent(body), base64 ? 'base64' : 'utf8');
+  } else {
+    const supportedSchemes = ['file', 'data'];
```
nit: hoist this to avoid re-instantiating the unchanging array

aduh95 replied:
It's the error case anyway, I think it's fine as is.

JakobJingleheimer commented on lib/internal/modules/esm/load.js:
```diff
@@ -69,12 +70,35 @@ async function getSource(url, context) {
+    const supportedSchemes = ['file', 'data'];
+    throw new ERR_UNSUPPORTED_ESM_URL_SCHEME(parsed, supportedSchemes);
+  }
+  if (policy?.manifest) {
+    policy.manifest.assertIntegrity(parsed, source);
+  }
```
```suggestion

  policy?.manifest.assertIntegrity(parsed, source);
```

(possibly might need another `?` between `manifest` and `assertIntegrity`)

aduh95 replied:
I copied that from defaultLoad, we should probably address both in a follow up PR.

JakobJingleheimer commented on lib/internal/modules/esm/load.js:
```diff
@@ -108,6 +132,51 @@ async function defaultLoad(url, context = kEmptyObject) {
 }
+/**
+ * @typedef LoadContext
+ * @property {string} [format] Hint possibly returned by `resolve`
```
```suggestion
 * @property {string} [format] A hint (possibly returned from `resolve`)
```

JakobJingleheimer commented on lib/internal/modules/esm/load.js:
```diff
@@ -108,6 +132,51 @@ async function defaultLoad(url, context = kEmptyObject) {
+/**
+ * @typedef LoadReturn
+ * @property {string} format format
+ * @property {string} responseURL url
```
```suggestion
 * @property {URL['href']} responseURL The module's fully resolved URL
```

JakobJingleheimer commented on lib/internal/modules/esm/load.js:
```diff
@@ -108,6 +132,51 @@ async function defaultLoad(url, context = kEmptyObject) {
+ */
+
+/**
+ * @param {string} url
```
```suggestion
 * @param {URL['href']} url
```

JakobJingleheimer commented on lib/internal/modules/esm/load.js:
```diff
@@ -108,6 +132,51 @@ async function defaultLoad(url, context = kEmptyObject) {
+ * @typedef LoadReturn
+ * @property {string} format format
+ * @property {string} responseURL url
+ * @property {Buffer} source source
```
I think `source` can be more than only a buffer, no?

aduh95 replied:
Yeah but `defaultLoad(Sync)` only returns `Buffer`s

JakobJingleheimer commented on lib/internal/modules/esm/load.js:
```diff
@@ -108,6 +132,51 @@ async function defaultLoad(url, context = kEmptyObject) {
+ */
+function defaultLoadSync(url, context = kEmptyObject) {
+  let responseURL = url;
+  const { importAssertions } = context;
+  let {
+    format,
+    source,
+  } = context;
+
+  format ??= defaultGetFormat(new URL(url), context);
```
Format does not / should not change, it just needs a default value, so I think it would be better to keep it a constant with a default:

```suggestion
  const {
    format = defaultGetFormat(new URL(url), context),
    importAssertions,
  } = context;

  let { source } = context;
```

aduh95 replied:
What if `format` is `null`?

JakobJingleheimer replied:
Ah, we have a bug in the `resolve` hook final result validation: https://github.com/nodejs/node/blob/d150316a8ecad1a9c20615ae62fcaf4f8d060dcc/lib/internal/modules/esm/hooks.js#L336-L339

`format` is an enum, so `null` is not valid (only the enum values or nothing)

Let's fix them together (we can't fix this part without creating a second bug).

aduh95 replied:
I'm pretty sure we've always accepted nullish values indifferently in the hook APIs.

JakobJingleheimer replied:
Even for enums? Why? `null` is an actual value.

JakobJingleheimer commented on lib/internal/modules/esm/loader.js:
```diff
@@ -388,6 +390,23 @@ class ModuleLoader {
+      this.#customizations.loadSync(url, context) :
+      defaultLoadSync(url, context);
+    let format = result?.format;
+    if (format === 'commonjs') {
+      format = 'require-commonjs';
+      result = { __proto__: result, format };
+    } else if (format === 'module') {
+      throw new ERR_REQUIRE_ESM(url, true);
+    }
```
nit: I think a switch is better here (more grokkable)

aduh95 replied:
```suggestion
    if (format === 'module') {
      throw new ERR_REQUIRE_ESM(url, true);
    }
    if (format === 'commonjs') {
      format = 'require-commonjs';
      result = { __proto__: result, format };
    }
```

JakobJingleheimer commented on lib/internal/modules/esm/module_job.js:
```diff
@@ -186,6 +195,20 @@ class ModuleJob {
+
+    this.module.instantiate();
+    this.instantiated = PromiseResolve();
+    const timeout = -1;
```
Just curious: is this merely to avoid a magic number?

aduh95 replied:
No idea, I've copied that from line 214.

JakobJingleheimer commented on lib/internal/modules/esm/translators.js:
```diff
@@ -140,35 +142,103 @@ function enrichCJSError(err, content, filename) {
+            break;
+          case '.node':
+            return CJSModule._load(specifier, module);
+          default:
```
nit: this looks like a mistake

```suggestion
          default: break;
```

aduh95 replied:
I'd argue that `default: break` looks even more like a mistake 😅 I'm going to add a comment instead.

JakobJingleheimer reviewed (approved):
LGTM. nicely succint 🙌

aduh95 commented on doc/api/esm.md:
```diff
@@ -945,20 +948,40 @@ validating the import assertion.
-overcome in the future.
+not possible to replace the value of a Node.js builtin (core) module.
+
+If a `load` hook returns commonjs `source`, that module's `require` calls will
+be processed by ESM loader with registered `resolve` and `load` hooks;
+`require.extensions` and monkey-patching on the CommonJS module loader will not
+apply. If `source` is undefined or `null`, it will be handled by the CommonJS
+module loader and not processed by registered hooks. This behavior for nullish
+`source` is temporary—in the future, nullish `source` will not be supported.
```
```suggestion
The value of `source` can be omitted for type `'commonjs'`. When a `source` is
provided, all `require` calls from this module will be processed by the ESM
loader with registered `resolve` and `load` hooks; all `require.resolve` calls
from this module will be processed by the ESM loader with registered `resolve`
hooks; `require.extensions` and monkey-patching on the CommonJS module loader
will not apply. If `source` is undefined or `null`, it will be handled by the
CommonJS module loader and `require`/`require.resolve` calls will not go through
the registered hooks. This behavior for nullish `source` is temporary — in the
future, nullish `source` will not be supported.
```

aduh95 commented on doc/api/esm.md:
```diff
@@ -945,20 +948,40 @@ validating the import assertion.
+module loader and not processed by registered hooks. This behavior for nullish
+`source` is temporary—in the future, nullish `source` will not be supported.
+
+The Node.js own `load` implementation, which is the value of `next` for the last
+loader in the `load` chain, returns `null` for `source` for backward
+compatibility. Here is an example loader that would opt-in to using the
+non-default behavior:
```
```suggestion
The Node.js own `load` implementation, which is the value of `next` for the last
loader in the `load` chain, returns `null` for `source` when `format` is
`'commonjs'` for backward compatibility. Here is an example loader that would
opt-in to using the non-default behavior:
```

aduh95 commented on lib/internal/modules/esm/loader.js:
```diff
@@ -388,6 +390,23 @@ class ModuleLoader {
+      this.#customizations.loadSync(url, context) :
+      defaultLoadSync(url, context);
+    let format = result?.format;
+    if (format === 'commonjs') {
+      format = 'require-commonjs';
+      result = { __proto__: result, format };
+    } else if (format === 'module') {
+      throw new ERR_REQUIRE_ESM(url, true);
+    }
```
```suggestion
    if (format === 'module') {
      throw new ERR_REQUIRE_ESM(url, true);
    }
    if (format === 'commonjs') {
      format = 'require-commonjs';
      result = { __proto__: result, format };
    }
```

aduh95 commented:
The added test started failing because another unrelated PR modified `test/common/index.js`. Since this is likely to happen again and be annoying, I modified the test to not count the require calls from that module.

GeoffreyBooth commented:
Getting lots of failures on the `test-esm-loader-hooks` test: https://ci.nodejs.org/job/node-test-commit-linux/53248/

Maybe that test needs more work? I assume you don’t need to rebase this off of `main` to get the fixes to that test that recently landed on `main`?

wrporter commented:
I do not understand the intricacies of this change, but it seems that it may have caused problems with how `esbuild` resolves dependencies. See https://github.com/evanw/esbuild/issues/587

I have a project where builds started failing on Node v20.6.0 and they work just fine on Node v20.5.1 using Vitest in a Nest.js application. The error looks like the following:

```
TypeError: Cannot set property Scope of #<Object> which has only a getter
❯ Object.<anonymous> node_modules/@nestjs/common/index.js:3:41
❯ Object.<anonymous> node_modules/@nestjs/common/pipes/parse-float.pipe.js:6:17
```

GeoffreyBooth commented:
It’s already been fixed on `main`: https://github.com/nodejs/node/pull/49500.
